Support live-reload of multi-project Gradle builds

garyttierney opened the pull request:
Extends the quarkusDev task to register all Java projects within the
Gradle workspace as modules with the DevModeContext, causing any source
changes within a Java file in the workspace to automatically recompile
and hot deploy.

This gets live-reloading working within a basic test project (like the one in https://github.com/quarkusio/quarkus-quickstarts/pull/229), but I think there might be a better approach to take here.

aloubyansky commented:
Thanks a lot for looking into this! We do need to watch only the app sources and the projects it depends on, not all the available sources. We've learnt that implementing that for Maven projects.

garyttierney commented:
> We do need to watch only the app sources and the projects it depends on, not all the available sources. We've learnt that implementing that for Maven projects.

Ah, I did wonder if that was a concern. Good to know, this seems achievable using the Gradle API so I'll respin a v2 of this patch.

garyttierney committed:
Support live-reload of multi-project Gradle builds

Extends the quarkusDev task to register all Java projects within the
Gradle workspace that the Quarkus application depends on as modules with
the DevModeContext, causing any source changes within a Java file in the
workspace to automatically recompile and hot deploy.

garyttierney commented:
@aloubyansky I've pushed a respin of this patch that only watches local projects that occurs in the dependency graph of the Quarkus project. I've tested it against a sample multi-module Gradle project and get the expected live reload behaviour.

I'm not sure how the Maven plugin handles it, but it seems like it'd be more correct to have some sort of "source based" `io.quarkus.bootstrap.model.AppArtifact` (or `io.quarkus.bootstrap.model.AppDependency`), but since these tasks are so inherently tied to the plugin implementations I'm not sure it's worth making the change.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:
```diff
@@ -237,6 +244,54 @@ public void startDev() {
+                }
+
+                Project dependencyProject = ((ProjectDependency) dependency).getDependencyProject();
+                localProjectDependencies.add(dependencyProject.getPath());
```
Having obtained `dependencyProject` is there a reason you are not processing it right away? Instead of collecting a set and then iterating through all the projects filtering out those that aren't deps?

garyttierney replied:
Doh, no reason for that. I originally didn't have a `Project` to work with, and after getting it from the `ProjectDependency` I forgot to factor out the first iteration that built the set.

My mistake, v3 incoming shortly!

aloubyansky commented:
> I'm not sure how the Maven plugin handles it, but it seems like it'd be more correct to have some sort of "source based" `io.quarkus.bootstrap.model.AppArtifact` (or `io.quarkus.bootstrap.model.AppDependency`), but since these tasks are so inherently tied to the plugin implementations I'm not sure it's worth making the change.

Right. It would make sense have a common API, especially if we could re-use a lot of what we do with it. However, the current priority is to get it working at least. Thanks a lot for your efforts so far!

aloubyansky reviewed (approved):
Great, thank you!
